README.md
# NAME

MojoX::CustomTemplateFileParser - Parses a custom Mojo template file format (deprecated)

![Requires Perl 5.14](https://img.shields.io/badge/perl-5.14-brightgreen.svg) [![Travis status](https://api.travis-ci.org/Csson/p5-mojox-customtemplatefileparser.svg?branch=master)](https://travis-ci.org/Csson/p5-mojox-customtemplatefileparser)

# VERSION

Version 0.1003, released 2015-11-26.

# SYNOPSIS

    use MojoX::CustomTemplateFileParser;

    my $parser = MojoX::CustomTemplateFileParser->new(path => '/path/to/file.mojo', output => [qw/Html Pod Test]);

    print $parser->to_html;
    print $parser->to_pod;
    print $parser->to_test;

# STATUS

Deprecated. Replaced by [Stenciller](https://metacpan.org/pod/Stenciller).

# DESCRIPTION

MojoX::CustomTemplateFileParser parses files containing [Mojo::Templates](https://metacpan.org/pod/Mojo::Template) mixed with the expected rendering.

The parsing creates a data structure that can be output in various formats using plugins.

Its purpose is to facilitate development of tag helpers.

## Options

**`path`**

The path to the file that should be parsed. Parsing occurs at object creation.

**`output`**

An array reference to plugins in the `::Plugin::To` namespace.

## Methods

No public methods. See plugins for output options.

# PLUGINS

Currently available plugins:

- [MojoX::CustomTemplateFileParser::To::Html](https://metacpan.org/pod/MojoX::CustomTemplateFileParser::To::Html)
- [MojoX::CustomTemplateFileParser::To::Pod](https://metacpan.org/pod/MojoX::CustomTemplateFileParser::To::Pod)
- [MojoX::CustomTemplateFileParser::To::Test](https://metacpan.org/pod/MojoX::CustomTemplateFileParser::To::Test)

# SEE ALSO

- [Dist::Zilla::Plugin::Test::CreateFromMojoTemplates](https://metacpan.org/pod/Dist::Zilla::Plugin::Test::CreateFromMojoTemplates)
- [Dist::Zilla::Plugin::InsertExample::FromMojoTemplates](https://metacpan.org/pod/Dist::Zilla::Plugin::InsertExample::FromMojoTemplates)

# SOURCE

[https://github.com/Csson/p5-mojox-customtemplatefileparser](https://github.com/Csson/p5-mojox-customtemplatefileparser)

# HOMEPAGE

[https://metacpan.org/release/MojoX-CustomTemplateFileParser](https://metacpan.org/release/MojoX-CustomTemplateFileParser)

# AUTHOR

Erik Carlsson <info@code301.com>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2015 by Erik Carlsson.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for MojoX-CustomTemplateFileParser

{{$NEXT}}
 [API Changes]

 [Bug Fixes]

 [Documentation]

 [Enhancements]

 [New Features]

 [Other Changes]

 [Requirements]

 [Testing]

0.1002 2015-11-26T16:48:07Z UTC
 - No changes.

0.1001 2015-11-26T16:44:23Z UTC
 - Deprecated.

0.1000 2014-09-10T21:42:50Z CEST
 - Bugfix: Lines meant to be parsed between test and expected could end up
   in expected
 - Bugfix: Malformed POD output near =begin/end html

0.09 2014-09-10T16:50:23Z CEST
 - Created a plugin architecture which allows for easy customization of
   output
 - Moved flatten() into ::Plugin::To::Test
 - Moved exemplify() into ::Plugin::To::Pod
 - Moved htmlify() into ::Plugin::To::Html
 - Removed parse(). Now parsing takes place at object creation
 - Changed to Moose, thanks to MooseX::Object::Pluggable

0.08 2014-09-09T23:28:44Z CEST
 - Add support for choosing to exemplify a test only if it is marked as an
   example in the source file
 - Fixed unitialized warning
 - Added htmlify() which turns the source file into html

0.07 2014-09-08T22:10:33Z CEST
 - Ensure adequate line breaks between templates and  expected (and other
   line types) in exemplify() output

0.06 2014-09-08T20:38:25Z CEST
 - Added test_index attribute for looking up tests by number
 - Added test_count() for looking up number of tests
 - Added exemplify() which translates a test number into a pod-able string

0.05 2014-09-08T12:45:36Z CEST
 - Fixed a bug where the last variable in a looped test sometimes were
   checked agains all expected results
 - Build fails on Travis for Perls < 5.14

0.04 2014-09-07T21:16:41Z CEST
 - Fixed a bug where the last test in a file sometimes were generated twice

0.03 2014-09-07T20:57:54Z CEST
 - Added support for ==test loop(var var2)== syntax to easily create
   similar tests where only one value is different

0.02 2014-09-07T16:42:25Z CEST
 - The test is skipped if the template part of a test is empty
 - Skip test by using ==no test==

0.01 2014-09-07T14:22:39Z CEST
 - Initial release



lib/MojoX/CustomTemplateFileParser/Plugin/To/Test.pm
package MojoX::CustomTemplateFileParser::Plugin::To::Test;

use strict;
use warnings;
use 5.10.1;

# VERSION
# ABSTRACT: Create tests

use Moose::Role;

sub to_test {
    my $self = shift;
    my $baseurl = $self->_get_baseurl;
    my $filename = $self->_get_filename;

    if(!scalar keys %{ $self->structure }) {
        $self->parse;
    }
    my $info = $self->structure;

    my @parsed = join "\n" => @{ $info->{'head_lines'} };

    TEST:
    foreach my $test (@{ $info->{'tests'} }) {
        next TEST if !scalar @{ $test->{'lines_template'} };

        my $expected_var = sprintf '$expected_%s%s' => $test->{'test_name'}, ($test->{'loop_variable'} ? "_$test->{'loop_variable'}" : '');

        push @parsed => "\n#** test from $filename, line $test->{'test_start_line'}" . ($test->{'loop_variable'} ? ", loop: $test->{'loop_variable'}" : '');
        push @parsed => sprintf 'my %s = qq{%s};' => $expected_var, join "\n" => @{ $test->{'lines_expected'} };

        push @parsed => sprintf q{get '/%s' => '%s';} => $test->{'test_name'}, $test->{'test_name'};
        push @parsed => sprintf q{$test->get_ok('/%s')->status_is(200)->trimmed_content_is(%s, '%s');}
                                => $test->{'test_name'}, $expected_var, sprintf qq{Matched trimmed content in $filename, line $test->{'test_start_line'}%s}
                                                                                => $test->{'loop_variable'} ? ", loop: $test->{'loop_variable'}" : '';
    }

    push @parsed => 'done_testing();';
    push @parsed => '__DATA__';


    foreach my $test (@{ $info->{'tests'} }) {
        next TEST if !scalar @{ $test->{'lines_template'} };

        push @parsed => sprintf '@@ %s.html.ep' => $test->{'test_name'};
        push @parsed => join "\n" => @{ $test->{'lines_template'} };
    }

    return join ("\n\n" => @parsed) . "\n";
}

1;

__END__

=pod

=head1 SYNOPSIS

  use MojoX::CustomTemplateFileParser;

  my $parser = MojoX::CustomTemplateFileParser->new(path => '/path/to/file.mojo', output => ['Test']);

  print $parser->to_test;

=head1 DESCRIPTION

MojoX::CustomTemplateFileParser::Plugin::To::Test is an output plugin to L<MojoX::CustomTemplateFileParser>.

=head2 to_test()

This method is added to L<MojoX::CustomTemplateFileParser> objects created with C<output =E<gt> ['Test']>.

It returns a string ready to be put in a L<Test::More> file.

=head1 SEE ALSO

=for :list
* L<Dist::Zilla::Plugin::Test::CreateFromMojoTemplates>
* L<MojoX::CustomTemplateFileParser::Plugin::To::Html>
* L<MojoX::CustomTemplateFileParser::Plugin::To::Pod>

=cut


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.042.
use strict;
use warnings;

use 5.014000;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Parses a custom Mojo template file format (deprecated)",
  "AUTHOR" => "Erik Carlsson <info\@code301.com>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "MojoX-CustomTemplateFileParser",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.014000",
  "NAME" => "MojoX::CustomTemplateFileParser",
  "PREREQ_PM" => {
    "HTML::Entities" => "3.65",
    "Moose" => "2.12",
    "MooseX::Object::Pluggable" => "0.0013",
    "Path::Tiny" => "0.050",
    "Storable" => "2.24"
  },
  "TEST_REQUIRES" => {
    "Data::Dump::Streamer" => "2.37",
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::Deep" => "0.110",
    "Test::More" => "0.96"
  },
  "VERSION" => "0.1003",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Data::Dump::Streamer" => "2.37",
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "HTML::Entities" => "3.65",
  "Moose" => "2.12",
  "MooseX::Object::Pluggable" => "0.0013",
  "Path::Tiny" => "0.050",
  "Storable" => "2.24",
  "Test::Deep" => "0.110",
  "Test::More" => "0.96"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/MojoX/CustomTemplateFileParser.pm
package MojoX::CustomTemplateFileParser;

use strict;
use warnings;
use 5.10.1;
# VERSION
# ABSTRACT: Parses a custom Mojo template file format (deprecated)

use Moose;
with 'MooseX::Object::Pluggable';

use HTML::Entities;
use Path::Tiny();
use Storable qw/dclone/;

has path => (
    is => 'ro',
    isa => 'Str',
    required => 1,
);
has structure => (
    is => 'rw',
    isa => 'HashRef',
    default => sub { { } },
);
has test_index => (
    is => 'rw',
    isa => 'HashRef',
    default => sub { { } },
);
has output => (
    is => 'ro',
    isa => 'ArrayRef',
    default => sub { [ ] },
);

sub BUILD {
    my $self = shift;
    $self->_parse;

    foreach my $plugin (@{ $self->output } ) {
        $self->load_plugin("To::$plugin");
    }
}

sub _parse {
    my $self = shift;
    my $baseurl = $self->_get_baseurl;
    my @lines = split /\n/ => Path::Tiny::path($self->path)->slurp;

    # matches ==test== ==no test== ==test loop(a thing or two)== ==test example ==test 1== ==test example 2==
    my $test_start = qr/==(?:(NO) )?TEST(?: loop\(([^)]+)\))?( EXAMPLE)?(?: (?:\d+))?==/i;
    my $template_separator = '--t--';
    my $expected_separator = '--e--';

    my $environment = 'head';

    my $info = {
        head_lines => [],
        tests      => [],
        indexed    => {},
    };
    my $test = {};

    my $row = 0;
    my $testcount = 0;

    LINE:
    foreach my $line (@lines) {
        ++$row;

        if($environment eq 'head') {
            if($line =~ $test_start) {

                my $skipit = $1;
                $test->{'loop'} = defined $2 ? [ split / / => $2 ] : [];
                $test = $self->_reset_test();

                if(defined $skipit && $skipit eq lc 'no') {
                    $test->{'skip'} = $skipit;
                }

                push @{ $info->{'head_lines'} } => '';
                $test->{'test_number'} = ++$testcount;
                $test->{'is_example'} = defined $3 ? 1 : 0;;
                $test->{'test_start_line'} = $row;
                $test->{'test_number'} = $testcount;
                $test->{'test_name'} = sprintf '%s_%s' => $baseurl, $testcount;
                $environment = 'beginning';

                next LINE;
            }
            push @{ $info->{'head_lines'} } => $line;
            next LINE;
        }
        if($environment eq 'beginning') {
            if($line eq $template_separator) {
                $environment = 'template';
                next LINE;
            }
            push @{ $test->{'lines_before'} } => $line;
            next LINE;
        }
        if($environment eq 'template') {
            if($line eq $template_separator) {
                if(scalar @{ $test->{'lines_template'} }) {
                    unshift @{ $test->{'lines_template'} } => '';
                    push @{ $test->{'lines_template'} } => '';
                }
                $environment = 'between';
                next LINE;
            }
            # If we have no template lines, don't push empty lines.
            # This way we can avoid empty templates, meaning we can leave empty test blocks in the
            # source files without messing up the tests.
            push @{ $test->{'lines_template'} } => $line if scalar @{ $test->{'lines_template'} } || $line !~ m{^\s*$};
            next LINE;
        }
        if($environment eq 'between') {
            if($line eq $expected_separator) {
                $environment = 'expected';
                next LINE;
            }
            push @{ $test->{'lines_between'} } => $line;
            next LINE;
        }
        if($environment eq 'expected') {
            if($line eq $expected_separator) {
                $environment = 'ending';
                if(scalar @{ $test->{'lines_expected'} }) {
                    unshift @{ $test->{'lines_expected'} } => '';
                    push @{ $test->{'lines_expected'} } => '';
                }
                next LINE;
            }
            push @{ $test->{'lines_expected'} } => $line;
            next LINE;
        }
        if($environment eq 'ending') {
            if($line =~ $test_start) {
                $self->_add_test($info, $test);

                $test = $self->_reset_test();
                my $skipit = $1;
                if(defined $skipit && $skipit eq lc 'no') {
                    $test->{'skip'} = 1;
                }
                $test->{'loop'} = defined $2 ? [ split / / => $2 ] : [];
                $test->{'test_start_line'} = $row;
                $test->{'test_number'} = ++$testcount;;
                $test->{'is_example'} = $3 || 0;
                $test->{'test_name'} = sprintf '%s_%s' => $baseurl, $testcount;
                $environment = 'beginning';

                next LINE;
            }
            push @{ $test->{'lines_after'} } => $line if scalar @{ $test->{'lines_after'} } || $line !~ m{^\s*$};
            next LINE;
        }
    }

    $self->_add_test($info, $test);

    $self->test_index(delete $info->{'indexed'});
    $self->structure($info);

    return $self;
}

sub test_count {
    my $self = shift;
    return keys %{ $self->{'test_index'} };
}

sub _add_test {
    my $self = shift;
    my $info = shift;
    my $test = shift;

    #* Nothing to test
    return if !scalar @{ $test->{'lines_template'} } || $test->{'skip'};

    #* No loop, just add it
    if(!scalar @{ $test->{'loop'} }) {
        push @{ $info->{'tests'} } => $test;
        $info->{'indexed'}{ $test->{'test_number'} } = [ $test ];
        return;
    }
    $info->{'indexed'}{ $test->{'test_number'} } = [ ];

    foreach my $var (@{ $test->{'loop'} }) {
        my $copy = dclone $test;

        map { $_ =~ s{\[var\]}{$var}g } @{ $copy->{'lines_template'} };
        map { $_ =~ s{\[var\]}{$var}g } @{ $copy->{'lines_expected'} };
        $copy->{'loop_variable'} = $var;
        $copy->{'test_name'} .= "_$var";
        push @{ $info->{'tests'} } => $copy;
        push @{ $info->{'indexed'}{ $copy->{'test_number'} } } => $copy;
    }
    return;

}

sub _reset_test {
    my $self = shift;
    return {
        is_example => 0,
        lines_before => [],
        lines_template => [],
        lines_after => [],
        lines_between => [],
        lines_expected => [],
        test_number => undef,
        test_start_line => undef,
        test_name => undef,
        loop => [],
        loop_variable => undef,
    };
}

sub _get_filename {
    return Path::Tiny::path(shift->path)->basename;
}

sub _get_baseurl {
    my $self = shift;
    my $filename = $self->_get_filename;
    (my $baseurl = $filename) =~ s{^([^\.]+)\..*}{$1}; # remove suffix
    $baseurl =~ s{-}{_};
    return $baseurl;
}

1;

__END__

=pod

=head1 SYNOPSIS

  use MojoX::CustomTemplateFileParser;

  my $parser = MojoX::CustomTemplateFileParser->new(path => '/path/to/file.mojo', output => [qw/Html Pod Test]);

  print $parser->to_html;
  print $parser->to_pod;
  print $parser->to_test;

=head1 STATUS

Deprecated. Replaced by L<Stenciller>.

=head1 DESCRIPTION

MojoX::CustomTemplateFileParser parses files containing L<Mojo::Templates|Mojo::Template> mixed with the expected rendering.

The parsing creates a data structure that can be output in various formats using plugins.

Its purpose is to facilitate development of tag helpers.

=head2 Options

B<C<path>>

The path to the file that should be parsed. Parsing occurs at object creation.

B<C<output>>

An array reference to plugins in the C<::Plugin::To> namespace.

=head2 Methods

No public methods. See plugins for output options.

=head1 PLUGINS

Currently available plugins:

=for :list
* L<MojoX::CustomTemplateFileParser::To::Html>
* L<MojoX::CustomTemplateFileParser::To::Pod>
* L<MojoX::CustomTemplateFileParser::To::Test>

=head1 SEE ALSO

=for :list
* L<Dist::Zilla::Plugin::Test::CreateFromMojoTemplates>
* L<Dist::Zilla::Plugin::InsertExample::FromMojoTemplates>

=cut


lib/MojoX/CustomTemplateFileParser/Plugin/To/Pod.pm
package MojoX::CustomTemplateFileParser::Plugin::To::Pod;

use strict;
use warnings;
use 5.10.1;

# VERSION
# ABSTRACT: Create pod

use Moose::Role;

sub to_pod {
    my $self = shift;
    my $test_index = shift;
    my $want_all_examples = shift || 0;

    my $tests_at_index = $self->test_index->{ $test_index };
    my @out = ();

    TEST:
    foreach my $test (@{ $tests_at_index }) {
        next TEST if $want_all_examples && !$test->{'is_example'};

        if(scalar @{ $test->{'lines_before'} }) {
            push @out => "\n", '=begin html', "\n",
                      '<p>', @{ $test->{'lines_before'} }, '</p>',
                      "\n", '=end html',
                      "\n";
        }

        push @out => @{ $test->{'lines_template'} }, "\n";
        if(scalar @{ $test->{'lines_between'} }) {
            push @out => '=begin html', "\n",
                      '<p>',@{ $test->{'lines_between'} }, '</p>',
                      "\n", '=end html',
                      "\n";
        }
        push @out => @{ $test->{'lines_expected' } }, "\n";
        if(scalar @{ $test->{'lines_after'} }) {
            push @out => '=begin html', "\n",
                      '<p>',@{ $test->{'lines_after'} }, '</p>',
                      "\n", '=end html', "\n";
        }
    }

    my $out = join "\n" => @out;
    $out =~ s{\n\n\n+}{\n\n}g;
    $out =~ s{\n+=begin html\n+}{\n\n=begin html\n\n}g;
    $out =~ s{\n+=end html\n+}{\n\n=end html\n\n}g;

    return $out;

}

1;

__END__

=pod

=head1 SYNOPSIS

  use MojoX::CustomTemplateFileParser;

  my $parser = MojoX::CustomTemplateFileParser->new(path => '/path/to/file.mojo', output => ['Pod']);

  print $parser->to_pod;

=head1 DESCRIPTION

MojoX::CustomTemplateFileParser::Plugin::To::Pod is an output plugin to L<MojoX::CustomTemplateFileParser>.

=head2 to_pod()

This method is added to L<MojoX::CustomTemplateFileParser> objects created with C<output =E<gt> ['Pod']>.

=head1 SEE ALSO

=for :list

* L<Dist::Zilla::Plugin::InsertExample::FromMojoTemplates>
* L<MojoX::CustomTemplateFileParser::Plugin::To::Pod>
* L<MojoX::CustomTemplateFileParser::Plugin::To::Test>

=cut


lib/MojoX/CustomTemplateFileParser/Plugin/ToHtml.pm
package MojoX::CustomTemplateFileParser::Plugin::ToTest;

use strict;
use warnings;
use 5.10.1;
our $VERSION = '0.08';

use Moose::Role;

sub to_html {
    my $self = shift;

    my @out = ();
    my $tests = $self->structure->{'tests'};

    foreach my $test (@{ $tests }) {
        push @out => (qq{<div class="panel panel-default"><div class="panel-body">});
        push @out => (@{ $test->{'lines_before'} }) if scalar @{ $test->{'lines_before'} };
        push @out => ('<pre>', HTML::Entities::encode_entities(join("\n" => @{ $test->{'lines_template'} })), '</pre>');
        push @out => (@{ $test->{'lines_between'} }) if scalar @{ $test->{'lines_between'} };
        push @out => ('<pre>', HTML::Entities::encode_entities(join("\n" => @{ $test->{'lines_expected'} })), '</pre>');
        push @out => (@{ $test->{'lines_after'} }, "<hr />") if scalar @{ $test->{'lines_after'} };
        push @out => (@{ $test->{'lines_expected'} });
        push @out => (qq{</div></div>});
    }

    return join '' => @out;
}

1;


lib/MojoX/CustomTemplateFileParser/Plugin/To/Html.pm
package MojoX::CustomTemplateFileParser::Plugin::To::Html;

use strict;
use warnings;
use 5.10.1;

# VERSION
# ABSTRACT: Create html

use Moose::Role;

sub to_html {
    my $self = shift;

    my @out = ();
    my $tests = $self->structure->{'tests'};

    foreach my $test (@{ $tests }) {
        push @out => (qq{<div class="panel panel-default"><div class="panel-body">});
        if(scalar @{ $test->{'lines_before'} }) {
            push @out => @{ $test->{'lines_before'} };
        }
        push @out => ('<pre>', HTML::Entities::encode_entities(join("\n" => @{ $test->{'lines_template'} })), '</pre>');
        if(scalar @{ $test->{'lines_between'} }) {
            push @out => @{ $test->{'lines_between'} };
        }
        push @out => ('<pre>', HTML::Entities::encode_entities(join("\n" => @{ $test->{'lines_expected'} })), '</pre>');
        if(scalar @{ $test->{'lines_after'} }) {
            push @out => @{ $test->{'lines_after'} };
        }
        push @out => '<hr />', @{ $test->{'lines_expected'} };
        push @out => (qq{</div></div>});
    }

    return join '' => @out;
}

1;

=pod

=head1 SYNOPSIS

  use MojoX::CustomTemplateFileParser;

  my $parser = MojoX::CustomTemplateFileParser->new(path => '/path/to/file.mojo', output => ['Html']);

  print $parser->to_html;

=head1 DESCRIPTION

MojoX::CustomTemplateFileParser::Plugin::To::Html is an output plugin to L<MojoX::CustomTemplateFileParser>.

=head2 to_html()

This method is added to L<MojoX::CustomTemplateFileParser> objects created with C<output =E<gt> ['Html']>.

=head1 SEE ALSO

=for :list
* L<Dist::Zilla::Plugin::InsertExample::FromMojoTemplates>
* L<MojoX::CustomTemplateFileParser::Plugin::To::Pod>
* L<MojoX::CustomTemplateFileParser::Plugin::To::Test>

=cut


lib/MojoX/CustomTemplateFileParser/Plugin/ToTest.pm
package MojoX::CustomTemplateFileParser::Plugin::ToTest;

use strict;
use warnings;
use 5.10.1;
our $VERSION = '0.08';

use Moose::Role;

sub to_test {
    my $self = shift;
    my $baseurl = $self->_get_baseurl;
    my $filename = $self->_get_filename;

    if(!scalar keys %{ $self->structure }) {
        $self->parse;
    }
    my $info = $self->structure;

    my @parsed = join "\n" => @{ $info->{'head_lines'} };

    TEST:
    foreach my $test (@{ $info->{'tests'} }) {
        next TEST if !scalar @{ $test->{'lines_template'} };

        my $expected_var = sprintf '$expected_%s%s' => $test->{'test_name'}, ($test->{'loop_variable'} ? "_$test->{'loop_variable'}" : '');

        push @parsed => "#** test from $filename, line $test->{'test_start_line'}" . ($test->{'loop_variable'} ? ", loop: $test->{'loop_variable'}" : '');
        push @parsed => sprintf 'my %s = qq{ %s };' => $expected_var, join "\n" => @{ $test->{'lines_expected'} };

        push @parsed => sprintf q{get '/%s' => '%s';} => $test->{'test_name'}, $test->{'test_name'};
        push @parsed => sprintf q{$test->get_ok('/%s')->status_is(200)->trimmed_content_is(%s, '%s');}
                                => $test->{'test_name'}, $expected_var, sprintf qq{Matched trimmed content in $filename, line $test->{'test_start_line'}%s}
                                                                                => $test->{'loop_variable'} ? ", loop: $test->{'loop_variable'}" : '';
    }

    push @parsed => 'done_testing();';
    push @parsed => '__DATA__';


    foreach my $test (@{ $info->{'tests'} }) {
        next TEST if !scalar @{ $test->{'lines_template'} };

        push @parsed => sprintf '@@ %s.html.ep' => $test->{'test_name'};
        push @parsed => join "\n" => @{ $test->{'lines_template'} };
    }

    return join ("\n\n" => @parsed) . "\n";
}

1;
